add suport for replay protection to TestingCommandFilter

andreimatei opened the pull request:
Turns out there can be raft replays in tests. This adds optional support for
replay protection in the mock by saving the raft cmd id in the context,
and uses it in the sql mocks.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5184)

<!-- Reviewable:end -->

andreimatei committed:
add suport for replay protection to TestingCommandFilter

Turns out there can be raft replays in tests. This adds optional support for
replay protection in the mock by saving the raft cmd id in the context,
and uses it in the sql mocks.

andreimatei commented:
notice @bdarnell

tschottdorf commented:
Reviewed 13 of 13 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/main_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcZMgzIgMQAFJwpMXF:-KCcZMgzIgMQAFJwpMXG:344456975) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/sql/main_test.go#L102)):*
I think this is over-engineered - simply make replay protection a flag on the `CommandFilters` struct whose effect is never giving a replay to the filter implementation. Populating a map will be cheap enough for any type of reasonable unit test.

---

*[storage/replica.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYZqSDxV0zzZuFUnl:-KCcYZqSDxV0zzZuFUnm:153375504) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L78)):*
the right way to do this is without an allocation every time is

```go
type raftCmdIDContextKeyType struct{}
var raftCmdIDContextKey raftCmdIDContextKeyType
```

---

*[storage/replica.go, line 1325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYsiTBr4cnDEkWfAh:-KCcYsiTBr4cnDEkWfAi:1915016339) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1325)):*
what is it with those `?!?`? This isn't a 14 year old's diary. :smile: 

---

*[storage/replica.go, line 1353 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCc_1gw9jeOxLqNdjoO:-KCc_1gw9jeOxLqNdjoP:1365690357) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1353)):*
do this only if a TestingCommandFilter is set.

---

*[storage/replica.go, line 1632 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYRhAelXNulJd8kYn:-KCcYRhAelXNulJd8kYo:-534665136) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1632)):*
This probably forces the index on the heap, so we allocate one item for each individual request in the batch. Make this conditional on having a TestingCommandFilter in place, and move the other WithValue next to it.

---

*[storage/replica.go, line 2036 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcaGPlX332NtC2vvTG:-KCcaGPmfJC4CF2F1JOv:1250296295) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L2036)):*
`r.context()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[sql/main_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcZMgzIgMQAFJwpMXF:-KCfQi0MTYgLXiMGkodF:1431721741) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/sql/main_test.go#L102)):*
I think I'd rather keep it like this; it's a better api forcing you to think about idempotency at the time you install a filter. Because otherwise we probably wouldn't set that flag by default, and then you'd forget to set it if you need it, etc. Also I don't think it'd save much code.

---

*[storage/replica.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYZqSDxV0zzZuFUnl:-KCfRiLNeG51x5VlBXcB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L78)):*
Done.

---

*[storage/replica.go, line 1325 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYsiTBr4cnDEkWfAh:-KCfRkDc2xF5Wduodxqu:1166285591) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1325)):*
SNAP! I was differentiating between an unexpected condition and an unrecoverable condition. But fine.

---

*[storage/replica.go, line 1353 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCc_1gw9jeOxLqNdjoO:-KCfSDvqVA9inFfPQZCp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1353)):*
Done.

---

*[storage/replica.go, line 1632 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYRhAelXNulJd8kYn:-KCfUzW5I6aTPIF--ox1:460555987) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1632)):*
Well I can't move the other WithValue, because I don't have a `cmdID` here, and the read call paths generally don't have such an id, right?
I've made this line conditional on TestingCommandFilter.
To avoid a new allocation here, I could put the whole struct{key, index} in the context in `processRaftCommand`, with a dummy index val, and the modify that struct here. But then I'd have to put a *struct in the context... I don't think it's worth it.

---

*[storage/replica.go, line 2036 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcaGPlX332NtC2vvTG:-KCfV93amlMX3Osm7MdB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L2036)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod making replay protection a bool on the `CommandFilter` struct.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[sql/main_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcZMgzIgMQAFJwpMXF:-KCgNJfyWpWjwx92CZmB:1088138977) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/sql/main_test.go#L102)):*
No, I was thinking to leave replay protection off by default. There's no indication of any other test except yours in the system that has replay protection flakiness (though it does happen occasionally), and the code in `*CommandFilters` is already a mouthful to read. KISS

---

*[storage/replica.go, line 78 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYZqSDxV0zzZuFUnl:-KCgO-DO5CQ6l2msFrDv:-1148703814) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L78)):*
I don't see it (or any other updates).

---

*[storage/replica.go, line 1632 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcYRhAelXNulJd8kYn:-KCgNl3bqMZekTJYJ1M9:28468781) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/storage/replica.go#L1632)):*
ah, yeah, no such thing on reads. With it being conditional on a mock set, allocs don't matter, so that's fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-:-KCgOArj3Wu0bAx2cp94:-478240665)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1310 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh96AKv9gOebZ6tNIO:-KCh96AKv9gOebZ6tNIP:-1286060156) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica.go#L1310)):*
Give this a different name (`RaftCmdIDAndIndex`?), because it's more than just the command ID.

---

*[storage/replica_command.go, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh8FuHCunJWJyHpDcp:-KCh8FuHCunJWJyHpDcq:241154898) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica_command.go#L62)):*
I'd rather make `commandID` and `index` arguments to the filter than pass them around through the `context.Context`. If you're concerned about the number of arguments and needing to touch every filter every time you change the signature you could put all the arguments into a `FilterContext` struct, so you can see what's there instead of the `context.Context` black box. 

Passing the `context.Context` around may be a good thing for other reasons (deadlines, cancellation), but passing necessary data through it is a bad idea.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/main_test.go, line 102 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCcZMgzIgMQAFJwpMXF:-KCqvSb3sKvS4eoS3fGk:1030777953) ([raw file](https://github.com/cockroachdb/cockroach/blob/95109093a462443576af25096e891db4ece0ff83/sql/main_test.go#L102)):*
i've extracted the replay protection into a CommandFilter wrapper that anyone can use. PTAL.

---

*[storage/replica.go, line 1310 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh96AKv9gOebZ6tNIO:-KCqvkZcA_OeG7Xu_U_c:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica.go#L1310)):*
Done.

---

*[storage/replica_command.go, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh8FuHCunJWJyHpDcp:-KCqvoNQmfNQyx_2e_Un:859143479) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica_command.go#L62)):*
ok, passing explicitly now.
Left the context too, seems like a good idea to pass it around.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/util/util.go, line 17 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr7SOsrx4KHfQE4Tpe:-KCr7SOt6za0sAtwnAwC:-1399599990) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/util/util.go#L17)):*
Why did this need to be extracted into a separate package? Do we ever use a command filter from a package that's not allowed to import `storage`?

The package should be named `storage/storageutil` (see `io/ioutil` in the standard library) so that you don't have to manually add an import alias everywhere you use it. 

---

*[storage/util/util.go, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr7meYMzkPMcWlDyE7:-KCr7meYMzkPMcWlDyE8:-1651060679) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/util/util.go#L23)):*
Why introduce a struct (with a name that precludes any future flexibility) instead of passing two separate arguments?

---

*[testutils/mocking.go, line 25 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr87AwHMgz_SAteUBC:-KCr87AwHMgz_SAteUBD:-195949496) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/mocking.go#L25)):*
The `testutils` package should have minimal dependencies; this should go into a subpackage like `testuils/gossiputil` and `testutils/sqlutils` (ugh, the inconsistent pluralization). Or I think it would be fine to put this in `storage` itself since that's where the command filter mechanism is defined.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 16 of 16 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1081 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrGuTClEv5_sK2gkFH:-KCrGuTClEv5_sK2gkFI:-1754238308) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/replica.go#L1081)):*
`storage_util` strikes me as an unidiomatic package name.

---

*[storage/replica_command.go, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh8FuHCunJWJyHpDcp:-KCrH6XwJ3t0IsG_gIAv:-1292041728) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica_command.go#L62)):*
Makes little sense to pass the context down into `executeCmd` if you ask me. Anything above is fine. We should generally not pass things around unless there's actually a reason to use it.

---

*[storage/util/util.go, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr7meYMzkPMcWlDyE7:-KCrHNf3XFzN7AI7yW8P:132116425) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/util/util.go#L23)):*
it's also only used during mocks - that's why I agreed to putting it in the context initially. It's awkward to plumb something down three levels just to use it in a mock. I'll defer to your judgment here though, somehow all options are a little unsavory.

---

*[testutils/mocking.go, line 28 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrHtLLVpVzIdXlBILn:-KCrHtLLVpVzIdXlBILo:168494337) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/mocking.go#L28)):*
Seems superfluous.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[testutils/mocking.go, line 28 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrHtLLVpVzIdXlBILn:-KCrXxwFvIEEZV7CsfE2:1330368954) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/mocking.go#L28)):*
I had to do it go get around a lint that wanted `error` to be the last thing returned from functions

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[testutils/mocking.go, line 28 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrHtLLVpVzIdXlBILn:-KCrkbmzZy9uO-b-iifO:1650705686) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/mocking.go#L28)):*
Where? I don't see that function. If it's required, add a comment explaining that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1081 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrGuTClEv5_sK2gkFH:-KD-qefx951cfET8i716:-911386752) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/replica.go#L1081)):*
changed to `storagebase`

---

*[storage/replica_command.go, line 62 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCh8FuHCunJWJyHpDcp:-KD-qkDneAxugFBYWP_b:-373330343) ([raw file](https://github.com/cockroachdb/cockroach/blob/38918d754c2beb4e7d6a3f8c7d51facd7ec8b6e1/storage/replica_command.go#L62)):*
I've hid it now in a `FilterArgs` struct. Doesn't hurt anything, and a filter sounds like the kind of thing that would like a `Context`...

---

*[testutils/storageutils/mocking.go, line 25 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr87AwHMgz_SAteUBC:-KD-qtjA1_9Zv6IF5-YP:610865803) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/storageutils/mocking.go#L25)):*
ok, created `testutils/storageutils`. I wonder if these all should have been `testutils/storagetestutils`

---

*[testutils/storageutils/mocking.go, line 28 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCrHtLLVpVzIdXlBILn:-KD-rN0b7D-yg7Z7JurY:-921755437) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/testutils/storageutils/mocking.go#L28)):*
Added a comment.

---

*[storage/util/util.go, line 17 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr7SOsrx4KHfQE4Tpe:-KCrZc5FKS3nIc6e_jKX:-2063112087) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/util/util.go#L17)):*
yeah, it's used in package sql, and sql can't import storage because, believe it or not, storage imports sql and breaking that was harder.

---

*[storage/util/util.go, line 23 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-KCr7meYMzkPMcWlDyE7:-KD-rPMXvsBUdS3mlM76:-2131266159) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae1edb04f9f3a22d52670c45cf32617e4f85fff5/storage/util/util.go#L23)):*
ok, made the struct private.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. I'm not sure that the `storagebase` package is worth it (the one type there could be replaced with a plain string) but there's been enough bikeshedding here and this looks fine.

---

Review status: 0 of 16 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5184#-:-KD5VoVJ8RmHfW4Buiox:59804056)*
<!-- Sent from Reviewable.io -->
